The Great JSON Escape

0xkarmacoma opened the pull request:
We now escape the `name` and `description` from an edition when we encode them in json+base64

This fixes the issue where an edition name/description contained reserved characters (most commonly `"`)

0xkarmacoma committed:
forge install foundry-rs/forge-std

0xkarmacoma committed:
Add Foundry EditionMetadataTest

0xkarmacoma committed:
Add .gas-snapshot

0xkarmacoma committed:
Fix bad json metadata when both imageUrl and animationUrl are set

0xkarmacoma committed:
Add test for name and description with reserved characters

Parsing fails with:
SyntaxError: Unexpected token e in JSON at position 13

0xkarmacoma committed:
Add testToken/ContractURIEscaped in EditionMetadataTest

0xkarmacoma committed:
Add basic escapeJsonString

Adapted from solidity-json-writer, but very inefficient

0xkarmacoma committed:
Update .gas-snapshot

0xkarmacoma committed:
Use solady's experimental new escapeJSON

Huge thanks to Vectorized 👌

testTokenURIEscaped() (gas: -81627 (-48.807%))
testContractURIEscaped() (gas: -81618 (-49.843%))

0xkarmacoma committed:
Using Base64 from Solady

0xkarmacoma committed:
Refresh dependencies

0xkarmacoma committed:
Add INTENSE tests

0xkarmacoma committed:
Update to most cutting edge LibString

0xkarmacoma committed:
Add the memory-safe tag to assembly in LibString

Otherwise compilation can fail with --via-ir:

YulException: Cannot swap Variable expr with Variable expr_1: too deep in the stack by 1 slots

0xkarmacoma committed:
Update gas snapshot

Vectorized commented:
Wow. LGTM.

Vectorized commented on contracts/EditionMetadataRenderer.sol:
```diff
@@ -99,14 +100,14 @@ contract EditionMetadataRenderer is EditionMetadataState {
                     // this is for opensea since they don't respect ERC2981 right now
-                    '", "seller_fee_basis_points": ',
+                    '","seller_fee_basis_points":',
                     StringsUpgradeable.toString(royaltyBPS),
```
Can use `LibString.toString` instead.

0xkarmacoma replied:
in [98d663a](https://github.com/showtime-xyz/nft-editions/pull/17/commits/98d663a712e3fdca82602b79cf5b0e35a22b3168)

Vectorized commented on contracts/EditionMetadataRenderer.sol:
```diff
@@ -99,14 +100,14 @@ contract EditionMetadataRenderer is EditionMetadataState {
                     StringsUpgradeable.toString(royaltyBPS),
-                    ', "fee_recipient": "',
+                    ',"fee_recipient":"',
                     StringsUpgradeable.toHexString(royaltyRecipient),
```
Can use `LibString.toHexString` instead.

0xkarmacoma replied:
in [98d663a](https://github.com/showtime-xyz/nft-editions/pull/17/commits/98d663a712e3fdca82602b79cf5b0e35a22b3168)

0xkarmacoma committed:
yarn upgrade ethereumjs-abi

Trying to get rid of the git+ssh dependency resolution that is causing errors on github actions

🙅‍♂️ resolved "git+ssh://git@github.com/ethereumjs/ethereumjs-abi.git#ee3994657fa7a427238e6ba92a84d0b529bbcde0"

0xkarmacoma committed:
Replace StringsUpgradeable uses with LibString

testLongEscapeControlChar() (gas: 0 (0.000%))
testLongEscapeDoubleQuotes() (gas: 0 (0.000%))
testEncodeContractURIIntenseDescription() (gas: -8564 (-0.012%))
testTokenURIEscaped() (gas: -2444 (-3.167%))
testTokenURI() (gas: -2444 (-3.757%))
testContractURIEscaped() (gas: -7312 (-9.957%))
testContractURI() (gas: -7484 (-12.150%))
Overall gas change: -28248 (-29.044%)

0xkarmacoma committed:
Drop the ?id=<tokenId> appended after media URLs

This makes the emitted json a little tighter and in line with the newer ZORA Drops (https://github.com/ourzora/zora-drops-contracts)

Test result: ok. 7 passed; 0 failed; finished in 399.84ms
testContractURI() (gas: 0 (0.000%))
testEncodeContractURIIntenseDescription() (gas: 0 (0.000%))
testLongEscapeControlChar() (gas: 0 (0.000%))
testLongEscapeDoubleQuotes() (gas: 0 (0.000%))
testContractURIEscaped() (gas: 33 (0.050%))
testTokenURIEscaped() (gas: -1501 (-2.009%))
testTokenURI() (gas: -1361 (-2.174%))
Overall gas change: -2829 (-4.133%)
